torchtitan/experiments/torchcomms/parallel_dims.py

```diff
@@ -134,25 +178,121 @@ def _build_mesh_without_ep(self) -> DeviceMesh:
+            dp_shard_in_ep = self.ep // (self.cp * self.tp)
+
+        mesh_shape = (
+            self.pp,
+            self.dp_replicate,
+            dp_shard_mod_ep,
+            dp_shard_in_ep,
+            self.cp,
+            self.tp,
+        )
+        mesh_dim_names = [
+            "pp",
+            "dp_replicate",
+            "dp_shard_mod_ep",
+            "dp_shard_in_ep",
+            "cp",
+            "tp",
+        ]
+
+        dims = [
+            d
+            for d, name in zip(mesh_shape, mesh_dim_names)
+            if d > 1 or name == "dp_shard_mod_ep"
+        ]
+        names = [
+            name
+            for d, name in zip(mesh_shape, mesh_dim_names)
+            if d > 1 or name == "dp_shard_mod_ep"
+        ]
 
+        logger.info(f"Building {len(dims)}-D device mesh with {names}, {dims}")
+
+        comm, device_mesh, mesh, cur_rank, comm_per_dim = _create_device_mesh(
+            self.world_size, mesh_shape, mesh_dim_names
+        )
+
+        if device_mesh is None:
+            return None
+
+        flatten_mesh = [
+            mesh.view(
+                self.pp,
+                self.dp_replicate * dp_shard_mod_ep * dp_shard_in_ep,
+                self.cp,
+                self.tp,
+            ),
+            mesh.view(
+                self.pp,
+                self.dp_replicate,
+                dp_shard_mod_ep * dp_shard_in_ep * self.cp,
+                self.tp,
+            ),
+            mesh.view(
+                self.pp,
+                self.dp_replicate * dp_shard_mod_ep * dp_shard_in_ep * self.cp,
+                self.tp,
+            ),
+            mesh.view(
+                self.pp,
+                self.dp_replicate,
+                dp_shard_mod_ep,
+                dp_shard_in_ep * self.cp * self.tp,
+            ),
+        ]
+
+        flattened_mesh_dim_names = ["dp", "dp_shard_cp", "dp_cp", "ep"]
+        flatten_mesh_dim_names = {
+            "dp": ["dp_replicate", "dp_shard_mod_ep", "dp_shard_in_ep"],
+            "dp_shard_cp": ["dp_shard_mod_ep", "dp_shard_in_ep", "cp"],
+            "dp_cp": ["dp_replicate", "dp_shard_mod_ep", "dp_shard_in_ep", "cp"],
+            "ep": ["dp_shard_in_ep", "cp", "tp"],
+        }
+
+        reshape_size = [
+            self.dp_replicate * dp_shard_mod_ep * dp_shard_in_ep,
+            dp_shard_mod_ep * dp_shard_in_ep * self.cp,
+            self.dp_replicate * dp_shard_mod_ep * dp_shard_in_ep * self.cp,
+            dp_shard_in_ep * self.cp * self.tp,
+        ]
+
+        flatten_ranks_per_dim = _calculate_ranks_per_dimension(
+            flatten_mesh, flattened_mesh_dim_names, reshape_size, cur_rank
+        )
+
+        _flatten_comms(
+            flatten_ranks_per_dim,
+            comm,
+            flatten_mesh_dim_names,
+            device_mesh,
+            comm_per_dim,
+        )
+
+        # call .finalize() later to release the sub comm before the root comm
```

Comment:
please add comment on where this is called? for example just say inside trainer.py? (Of course, with details)